src/hooks.ts

```diff
@@ -36,3 +42,113 @@ export const useHasResourcePermission = (resource: Resource, authzUrl: string, p
     const { permissions, isFetching } = useResourcePermissions(resource, authzUrl) ?? {};
     return { isFetching, hasPermission: permissions.includes(permission) };
 };
+
+export const useOpenIdConfig = (openIdConfigUrl: string) => {
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        dispatch(fetchOpenIdConfiguration(openIdConfigUrl));
+    }, [dispatch, openIdConfigUrl]);
+
+    return useSelector((state: RootState) => state.openIdConfiguration.data);
+}
+
+export const useSignInPopupTokenHandoff = (
+    applicationUrl: string,
+    authCallbackUrl: string,
+    clientId: string,
+    windowMessageHandler: MutableRefObject<null | MessageHandlerFunc>
+) => {
+    const dispatch = useDispatch();
+    const removeListenerIfNeeded = () => {
+        if (windowMessageHandler.current) {
+            window.removeEventListener("message", windowMessageHandler.current);
+        }
+    }
+    useEffect(() => {
+        removeListenerIfNeeded();
+        windowMessageHandler.current = (e: MessageEvent) => {
+            if (e.origin !== applicationUrl) return;
+            if (e.data?.type !== AUTH_RESULT_TYPE) return;
+            const { code, verifier } = e.data ?? {};
+            if (!code || !verifier) return;
+            localStorage.removeItem(LS_SIGN_IN_POPUP);
+            dispatch(tokenHandoff({ code, verifier, clientId: clientId, authCallbackUrl: authCallbackUrl }));
+        };
+        window.addEventListener("message", windowMessageHandler.current);
+
+        // Listener cleanup
+        return removeListenerIfNeeded;
+    }, [dispatch, removeListenerIfNeeded, applicationUrl, authCallbackUrl, clientId]);
```

Comment:
removeListenerIfNeeded should not be a dependency; it should be an anonymous function returned above